python/semantic_kernel/plugin_definition/kernel_plugin_collection.py

```diff
@@ -0,0 +1,225 @@
+
+        Raises:
+            ValueError: If the function or plugin_name is None or invalid.
+        """
+        if not function or not plugin_name:
+            raise ValueError("Function and plugin_name must not be None or empty")
+        if plugin_name in self.plugins:
+            raise ValueError(f"Plugin with name {plugin_name} already exists")
+
+        plugin = DefaultKernelPlugin.from_function(function)
+        self.plugins[plugin_name] = plugin
+
+    def add_functions_to_plugin(self, functions: List["KernelFunctionBase"], plugin_name: str) -> None:
+        """
+        Add a function to a plugin in the collection
+
+        Args:
+            functions (List[KernelFunctionBase]): The function to add to the plugin.
+            plugin_name (str): The name of the plugin to add the function to.
+
+        Raises:
+            ValueError: If the function or plugin_name is None or invalid.
+            ValueError: if the function already exists in the plugin.
+        """
+        if not functions or not plugin_name:
+            raise ValueError("Functions and plugin_name must not be None or empty")
+        if plugin_name not in self.plugins:
+            self.plugins.add(DefaultKernelPlugin(name=plugin_name))
+
+        plugin = self.plugins[plugin_name]
+        for func in functions:
+            if func.name not in plugin.functions:
+                plugin.functions[func.name] = func
+            else:
+                raise ValueError(f"Function with name '{func.name}' already exists in plugin '{plugin_name}'")
+
+    def add_range(self, plugins: List[KernelPlugin]) -> None:
+        """
+        Add a list of plugins to the collection
+
+        Args:
+            plugins (List[KernelPlugin]): The plugins to add to the collection.
+
+        Raises:
+            ValueError: If the plugins list is None.
+        """
+
+        if plugins is None:
+            raise ValueError("Plugins must not be None")
+        for plugin in plugins:
+            self.add(plugin)
+
+    def remove(self, plugin: KernelPlugin) -> bool:
+        """
+        Remove a plugin from the collection
+
+        Args:
+            plugin (KernelPlugin): The plugin to remove from the collection.
+
+        Returns:
+            True if the plugin was removed, False otherwise.
+        """
+        if plugin is None or plugin.name is None:
+            return False
+        return self.plugins.pop(plugin.name, None) is not None
+
+    def remove_by_name(self, plugin_name: str) -> bool:
+        """
+        Remove a plugin from the collection by name
+
+        Args:
+            plugin_name (str): The name of the plugin to remove from the collection.
+
+        Returns:
+            True if the plugin was removed, False otherwise.
+        """
+        if plugin_name is None:
+            return False
+        return self.plugins.pop(plugin_name, None) is not None
+
+    def get_plugin(self, name: str) -> Optional[KernelPlugin]:
```

Comment:
should we add a get_function(self, plugin_name: str, function_name: str) here? would make looking up functions for auto-executing tool_calls much easier and cleaner...

Reply:
Is the method chaining like: `plugin_collection.get_plugin(plugin_name).get_function(function_name)` not adequate? You'd rather see `plugin_collection.get_function(plugin_name, function_name`?

Reply:
I've added this.